cpp/include/raft/lap/lap_kernels.cuh

```diff
@@ -0,0 +1,502 @@
+    if (slack < d_col_slacks[COLID]) {
+      d_col_slacks[COLID] = slack;
+      d_col_parents[COLID] = ROWID;
+    }
+
+    if (near_zero(d_col_slacks[COLID])) {
+      if (nxt_rowid != -1) {
+        d_row_parents[NXT_ROWID] = COLID;  // update parent info
+
+        d_row_covers[NXT_ROWID] = 0;
+        d_col_covers[COLID] = 1;
+
+        if (d_row_visited[NXT_ROWID] != VISITED)
+          d_row_visited[NXT_ROWID] = ACTIVE;
+      } else {
+        d_col_visited[COLID] = REVERSE;
+        *d_flag = true;
+      }
+    }
+  }
+  d_row_visited[ROWID] = VISITED;
+}
+
+// Device function for traversing an alternating path from unassigned row to unassigned column.
+template <typename vertex_t>
+__device__ void __reverse_traversal(
+  int *d_row_visited, vertex_t *d_row_children, vertex_t *d_col_children,
+  vertex_t const *d_row_parents, vertex_t const *d_col_parents, int cur_colid) {
+  int cur_rowid = -1;
+
+  while (cur_colid != -1) {
+    d_col_children[cur_colid] = cur_rowid;
+    cur_rowid = d_col_parents[cur_colid];
+
+    d_row_children[cur_rowid] = cur_colid;
+    cur_colid = d_row_parents[cur_rowid];
+  }
+  d_row_visited[cur_rowid] = AUGMENT;
+}
+
+// Device function for augmenting the alternating path from unassigned column to unassigned row.
+template <typename vertex_t>
+__device__ void __augment(vertex_t *d_row_assignments,
+                          vertex_t *d_col_assignments,
+                          vertex_t const *d_row_children,
+                          vertex_t const *d_col_children, vertex_t cur_rowid,
+                          vertex_t N) {
+  int cur_colid = -1;
+
+  while (cur_rowid != -1) {
+    cur_colid = d_row_children[cur_rowid];
+
+    d_row_assignments[cur_rowid] = cur_colid % N;
+    d_col_assignments[cur_colid] = cur_rowid % N;
+
+    cur_rowid = d_col_children[cur_colid];
+  }
+}
+
+// Kernel for reducing the rows by subtracting row minimum from each row element.
+//  FIXME:  Once cuda 10.2 is the standard should replace passing infinity
+//          here with using cuda::std::numeric_limits<weight_t>::max()
+template <typename vertex_t, typename weight_t>
+__global__ void kernel_rowReduction(weight_t const *d_costs,
+                                    weight_t *d_row_duals, int SP, vertex_t N,
+                                    weight_t infinity) {
+  int spid = blockIdx.y * blockDim.y + threadIdx.y;
+  int rowid = blockIdx.x * blockDim.x + threadIdx.x;
+  weight_t min = infinity;
+
+  if (spid < SP && rowid < N) {
+    for (int colid = 0; colid < N; colid++) {
+      weight_t slack = d_costs[spid * N * N + rowid * N + colid];
+
+      if (slack < min) {
+        min = slack;
+      }
+    }
+
+    d_row_duals[spid * N + rowid] = min;
+  }
+}
+
+// Kernel for reducing the column by subtracting column minimum from each column element.
+//  FIXME:  Once cuda 10.2 is the standard should replace passing infinity
+//          here with using cuda::std::numeric_limits<weight_t>::max()
```

Comment:
might be a good idea to open a github issue to track

Reply:
Created #96